Dshot bidir blhelis support

joelucid opened the pull request:
This PR adds support for a modified dshot bidir protocol which can be supported by blheli_s hw. The main changes are (a) erpm periods instead of frequencies are reported to avoid the need for a division on efm8bb2 hw and (b) a line protocol with lower resolution is used for the backchannel, allowing blheli_s hw to bitbang the protocol since no additional hw timers are available.

The type of protocol is auto-detected using the number of level changes in the packet.

Here's a flight demo of rpm filter running on a blheli_s esc: https://youtu.be/6XwTDUYOJfY

joelucid commented:
Ok - this overflows iTCM RAM on f7. Probably worth to have the code path in ITCM RAM though since its executed for every motor. @mikeller any idea where we might gain a bit of space?

mikeller commented:
I don't think that starting to introduce new protocols that are to be used by proprietary software is the right way to go for this - we have had this with ProShot, and it's considerably increased code size and complexity.
I'd prefer if we changed the standard for Dshot RPM feedback to have this new format for all ESC firmwares supporting it.
Since we will have to make a breaking change for all ESC firmware supporting RPM feedback to switch to inverted Dshot, including this change in it as well will work without any additional disruption for users.
@joelucid, @sskaug, @ronlix: Thoughts?

joelucid commented:
Yes, @mikeller, I agree it make sense to consider using the new protocol for signal line rpm telemetry in general. The main reason is that it will make gpio bit banging the protocol much more viable (see https://github.com/betaflight/betaflight/pull/7446 for discussion of the bitbang issue and the GCR protocol, @sskaug & @ronlix).

Of course there's a bit of an ugly asymmetry there if we use regular dshot for the uplink and the modified line coding for the downlink. But this actually matches our requirements fairly well: bit banging the new line coding for the downlink at the FC will require the same resolution as the existing dshot line coding for the uplink (3x bitrate). It's the input side that causes the problems as long as output is DMA driven. Only on Blheli_S is the output side also a problem since it has to be bit banged by the cpu. And decoding regular dshot requires less code than the new line coding - so the weaker ESC hw benefits from receiving regular dshot.

My attempts at using proshot have never really resulted in happiness. The long pulses make the protocol very susceptible to synchronization issues. It's the exact opposite of GCR in that sense. So one might also consider removing proshot to keep the number of supported protocols manageable.

sskaug commented:
@joelucid Can you explain a bit more about what you are proposing here? Is it GCR encoded erpm period for the return data? 

Have anyone tried to implement this in BLHeli_S?

I totally agree on Proshot. It is very susceptible to timing errors and will not be reliable on the ESC hardwares out there. It could preferably be removed to reduce code size and verification efforts.

joelucid commented:
> @joelucid Can you explain a bit more about what you are proposing here? Is it GCR encoded erpm period for the return data?
> Have anyone tried to implement this in BLHeli_S?

Yes, @sskaug, I've been testing code reporting the eperiod using the following encoding: eperiod is reported in 1us units in an exponential format with a 3 bit exponent and 9 bit mantisse (eeem mmmm mmmm) for a range from 1 * 2^0 * 1us = 1us  to 511 * 2 ^ 7 = 65408 * 1us =  0.065408s. This value sits in the upper 12 bits of the packet and an inverted xor of its nibbles in the check sum as with inverted dshot. This value is then GCR encoded to achieve a line coding with fewer level transitions.

I've successfully tested this scheme on BLHeli_S hw with dshot 300 for the uplink. I'm using a 500kbaud bitrate for the backchannel. Due to GCR a telemetry packet has 20 bits which means a frame length of 20 / 500kbaud = 40us vs 53us for a regular dshot packet.

mikeller commented:
Having a look at the statistics from last month, I think that ditching ProShot is definitely worth consideration:

![image](https://user-images.githubusercontent.com/4742747/61089381-fbe65b80-a48e-11e9-9ba2-2772eb0251ba.png)

But this is a matter that is independent of the discussion we are having here about changing the protocol for Dshot RPM feedback.

sskaug commented:
I am a bit surprised that this comes up now. I thought our plan was to ensure stability of the current implementation (inverted dshot / test code 32.6.5), and then release this as a full release. 32.6.5 has now been out for about a month, and to my knowledge there are no reported issues. So we should now, (or very soon), have the required confidence to do a full release of this code.

What is the purpose of doing GCR on the return path?
- Is it to reduce the resource usage in the FC? I thought the inverted dshot return path code now runs acceptably on all BF targets?
- Is it to ease implementation of BLHeli_S?

mikeller commented:
@sskaug: Sorry for that, and yes, that was the plan a couple of months back.
I think the fact that we now have a workable implementation of Dshot RPM feedback for BLHeli_S, and that this implementation requires some optimisations to make it feasible on BLHeli_S, are reason enough to re-evaluate the original plan.
If the choice is between a protocol that is supportable by closed source firmwares on two different hardware types, and one that is supportable by closed source firmwares on three different hardware types, then I think the better choice is for Betaflight to go for the protocol that is supported on more hardware types.
Happy to help with the implementation of the new feedback protocol, and I think @joelucid will be so as well, so let us know if we can help.

joelucid commented:
> What is the purpose of doing GCR on the return path?
> Is it to reduce the resource usage in the FC? I thought the inverted dshot return path code now runs acceptably on all BF targets?
> Is it to ease implementation of BLHeli_S?

@sskaug, the current implementation does run well on most BF targets but it requires reconfiguring some dma channels for certain targets which impairs other functionality like LEDs. There are some boards that don't work without hw mods since the motor pins can't be used as input for a timer.

We have an idea for an implementation that would fix all this by bit banging using gpio dma. But this is not viable with dshot due to the oversampling required for input. GCR would make it workable.

Then secondly - and this is the reason for this PR - GCR makes an rpm telemetry implementation possible on BLHeli_S ESCs where the protocol needs to be bit banged, too, and the hw just can't do it reliably using plain dshot 300.

Personally I don't see an overwhelming case to force blheli32 to use this GCR protocol. It would make it future proof for a bit banging gpio implementation on the FC though. And it would eliminate code on the FC which might fix the ITCM RAM overflow and simplify things overall.

sskaug commented:
Thanks for the info! Implementing GCR coded return data for BLHeli_32 is surely easy.
How is the GCR encoding and transmission done in BLHeli_S, is it done on the main thread, or are interrupts used? Running 40us on the main thread will surely degrade performance at higher rpms. Also is GPIO bit banging proven to work for CGR, or can there still be some issues.
What is the amount of test data on BLHeli_S? And GPIO bit banging? I think it would be good to have at least one of these proven with some volume testing before commiting to this change.

joelucid commented:
@sskaug, for the implementation on blheli_s hw I've modified blheli_s to support running optional add-on services. The telemetry service I implemented for rpm telemetry runs in the background, encoding telemetry frames to be sent when the main esc code doesn't need the cpu. If such a telemetry frame is available when the esc receives the dshot frame it is sent using timer 4 interrupts. I only support the efm8bb21 where timer 4 is available and unused. The implementation has been tested internally on maybe 80 quads with no issues - even tiny high rpm quads. We've seen no impact on max attainable rpm.

When using dshot 300 proper for the backchannel I would get corrupted packets at higher rpm due to the jitter introduced by the variability of interrupt latency inherent in the 8051 design and aggravated by the various interrupt disabled critical sections in the code. GCR solved this completely due to its great jitter robustness.

GPIO bitbanging has been prototyped for output using dshot, but we haven't tested the input side yet. But I have a strong sense that dma sampling gpio at 3x the shortest cycle time of the sender will be sufficient to decode the GCR frames with their max of 3 consecutive same output levels. This is likely true for dshot, too - but there the shortest cycle time is 3x the bitrate, so you'd have to sample at 9x bitrate which is prohibitively expensive.

joelucid commented:
Here's an example of the code in action on a blheli_s style esc: https://youtu.be/6XwTDUYOJfY

joelucid commented:
@sskaug did some test with dshot600 today or rather for the backchannel for it. Tried GCR at 750kbit and got it working with error rates < 0.2%. I think this is a change that might make sense now that we consider this for the general solution: GCR uses 5/4th of the bits so the bitrate should be 5/4th of the dshot bitrate. Let me know what you think.

ledvinap commented:
A side note: I didn't analyze it too far, but is seems that 1.5x oversampling (1.5 synchronous samples per bit period) should be enough to handle NRZ protocol including clock drift...

joelucid commented:
Got it working with dshot 600:

![IMG_1994](https://user-images.githubusercontent.com/18134492/61394573-b88f5100-a8c3-11e9-8a66-38915b0dbf94.jpg)

@sskaug, as discussed above the GCR bitrate is now defined as 5/4 * bitrate of the dshot transfer rate.

joelucid commented:
Let me specify the new rpm telemetry protocol here for you @sskaug, and others:

Dshot bidir uses inverted signal levels (idle is 1). FC to ESC uses dshot frames but the lowest 4 bits hold the complement of the other nibbles xor'd together (normal dshot does not complement the xor sum). The ESC detects based on the inversion that telemetry packets have to be sent. 

30us after receiving the dshot frame the esc responds with a telemetry frame. Logically the telemetry frame is a 16 bit value and the lowest 4 bits hold the complemented xor'd nibbles again. 

The upper 12 bit contain the eperiod (1/erps) in the following bitwise encoding:

```
e e e m m m m m m m m m
```

The 9 bit value M needs to shifted left (E - 1) times to get the period in micro seconds. This gives a range of 0.5 us to 32704 us. Which translates to a min e-frequency of 30.6 hz or for 14 pole motors 4.36 hz.

This 16 bit value is then GCR encoded to a 20 bit value by applying the following map nibble-wise:
```
0 -> 19
1 -> 1b
2 -> 12
3 -> 13
4 -> 1d
5 -> 15
6 -> 16
7 -> 17
8 -> 1a
9 -> 09
a -> 0a
b -> 0b
c -> 1e
d -> 0d
e -> 0e
f -> 0f 
```

This creates a 20 bit value which has no more than two consecutive zeros. This value is mapped to a new 21 bit value by starting with a bit value of ``0`` and changing the bit value in the next bit if the current bit in the incoming value is a ``1``, but repeating the previous bit value otherwise. Example:

``1 0 1 1 1 0 0 1 1 0`` would become ``0 1 1 0 1 0 0 0 1 0 0``.

This 21 bit value is then sent uninverted at a bitrate of 5/4 * the dshot bitrate. So Dshot 3 uses 375 kbit, dshot 600 uses 750 kbit. 

The esc needs to be ready after 2 * 30us + 2 * dshot frame length to receive the next dshot packet.

See https://en.wikipedia.org/wiki/Run-length_limited#GCR:_(0,2)_RLL for more details on the GCR encoding.

sskaug commented:
So now I assume we will only have the above signaling spec for realtime erpm telemetry.
What about Proshot? Will Proshot support be removed from BF? Or is it assumed that inverted Proshot shall also respond with GCR encoded erpm telemetry?

joelucid commented:
My code will assume that proshot telemetry also sends GCR. I’m removing everything other than GCR for the back channel. Keeping things simple. Rg removal of Proshot overall no call has been made yet.

> On Jul 19, 2019, at 3:18 PM, sskaug <notifications@github.com> wrote:
> 
> So now I assume we will only have the above signaling spec for realtime erpm telemetry.
> What about Proshot? Will Proshot support be removed from BF? Or is it assumed that inverted Proshot shall also respond with GCR encoded erpm telemetry?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/8554?email_source=notifications&email_token=AEKLLXBY42BXRQ6GIMHCRMLQAG5MDA5CNFSM4H7M2XZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2LTEJA#issuecomment-513225252>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEKLLXG3YSHCUMNGEWJML2DQAG5MDANCNFSM4H7M2XZA>.
>

joelucid commented:
@mikeller we now have a version of blheli32 which supports GCR. Rebased.

mikeller commented on src/main/drivers/dshot_dpwm.h:
```diff
@@ -55,16 +59,19 @@ motorDevice_t *dshotPwmDevInit(const struct motorDevConfig_s *motorConfig, uint1
 #define DSHOT_DMA_BUFFER_ATTRIBUTE FAST_RAM_ZERO_INIT
 #else
-#define DSHOT_DMA_BUFFER_ATTRIBUTE // None
+#ifdef STM32F7
```
This will never be true.

joelucid replied:
Good catch - this was from the last rebase since jflyper has now fixed it

mikeller commented on src/main/drivers/pwm_output_dshot.c:
```diff
@@ -101,7 +93,12 @@ FAST_CODE void pwmDshotSetDirectionOutput(
         motor->isInput = true;
-        motor->timer->inputDirectionStampUs = micros();
+        if (!inputStamp) {
+            inputStamp = micros();
```
Common practice in Betaflight is to use the suffix `...Us` for microsecond counts.

mikeller commented on src/main/drivers/pwm_output_dshot_hal.c:
```diff
@@ -148,26 +136,26 @@ FAST_CODE void pwmCompleteDshotMotorUpdate(void)
             LL_EX_TIM_EnableIT(dmaMotorTimers[i].timer, dmaMotorTimers[i].timerDmaSources);
             dmaMotorTimers[i].timerDmaSources = 0;
-        }
+       }
```
:space_invader:

joelucid replied:
fixed

mikeller commented on src/main/drivers/pwm_output_dshot_hal.c:
```diff
@@ -148,26 +136,26 @@ FAST_CODE void pwmCompleteDshotMotorUpdate(void)
-        if (motor->isInput) {
-            processInputIrq(motor);
-        } else
+            uint32_t irqStart = micros();
```
`timeUs_t irqStartUs`

joelucid replied:
fixed

mikeller commented on src/main/drivers/pwm_output_dshot_shared.c:
```diff
@@ -152,56 +153,65 @@ FAST_CODE void pwmWriteDshotInt(uint8_t index, uint16_t value)
-        if (diff > 0) {
-            if (diff >= 11) value |= 1;
+    uint32_t oldValue = buffer[0];
+    int      bits = 0;
```
Extra whitespace. :space_invader:

joelucid replied:
fixed

mikeller commented on src/main/io/serial_4way.c:
```diff
@@ -76,9 +79,9 @@
 #define SERIAL_4WAY_VER_MAIN 20
 #define SERIAL_4WAY_VER_SUB_1 (uint8_t) 0
-#define SERIAL_4WAY_VER_SUB_2 (uint8_t) 03
+#define SERIAL_4WAY_VER_SUB_2 (uint8_t) 04
```
Not sure how this is related to the bidirectional Dshot changes. Should this go into a separate pull request?

joelucid replied:
This is required for the extension to the 4way esc protocol which allows resetting the esc. I need to reset the esc to install jesc so it is related to dshot bidirectional support on blhelis hw. I could still move it into a different PR if you think that would be beneficial.

mikeller replied:
Yes please - this could potentially be contentious if it turns out to create issues with other hardware using the 4way protocol, so having it isolated will make resolving this easier.

McGiverGim commented:
Link to the blheli32 version, if someone wants to test it: https://github.com/bitdump/BLHeli/tree/master/BLHeli_32%20ARM/BLHeli_32%20Test%20code%20Rev32.6.4%20hex%20files/Rev32.6.6%20for%20inverted%20bidirectional%20dshot%20and%20GCR%20return%20data%20only

mikeller commented on src/main/drivers/pwm_output_dshot_shared.c:
```diff
@@ -241,64 +251,65 @@ void updateDshotTelemetryQuality(dshotTelemetryQuality_t *qualityStats, bool pac
 #endif // USE_DSHOT_TELEMETRY_STATS
 
-bool pwmStartDshotMotorUpdate(void)
+NOINLINE bool pwmStartDshotMotorUpdate(void)
```
This should be `FAST_CODE_NOINLINE`.

hydra commented:
Just so people following this PR, it broke H7 DSHOT.

@joelucid @jflyper and I have had conversations about this in slack, a new issue was created to track the issue here: https://github.com/betaflight/betaflight/issues/8706

Just as another quick reminder can those authoring/reviewing/approving be mindful that H7 support should be done at the same time as other MCUs.

joelucid commented:
Just so people are aware this PR addressed two issues which the motor refactor introduced: 1st it fixed the broken f7 dshot support and 2nd it fixed that dshot would not get reenabled after exiting the blhelis configurator. Unfortunately @hydra is right, h7 broke due to the fix of the second problem.

Now that h7 can be reliably built it'll be easier to avoid breaking it. Thanks @hydra!

Igor-Misic commented:
Can somebody explain the return signal from ESC to FC? Are bits represent as PWM like in DShot or it is more like a signal in UART? I can understand the logic from this post https://github.com/betaflight/betaflight/pull/8554#issuecomment-512507625

but it is not clear how data is sent on the physical layer. 

![image](https://user-images.githubusercontent.com/10188706/68023328-2b6fdc80-fca7-11e9-8d1a-56e36402436e.png)

joelucid commented:
The comment you referenced does describe the physical layer in the last step. The last bit sequence is just put to the wire.

markusgritsch commented:
@joelucid: From the RPM telemetry protocol specification:

> This 21 bit value is then sent uninverted at a bitrate of 5/4 * the dshot bitrate. So Dshot 3 uses 375 kbit, dshot 600 uses 750 kbit.

Are these bitrates of the telemetry packet still valid, or have they been changed for the final implementation?

joelucid commented:
That’s still correct, @markusgritsch.

jimkazmer commented:
Is the information described here (https://github.com/betaflight/betaflight/pull/8554#issuecomment-512507625) still accurate/current?

If it is, I need help understanding the reverse transformation of: 
> changing the bit value in the next bit if the current bit in the incoming value is a `1`, but repeating the previous bit value otherwise.
I read through the code decode_bb() and I didn't understand how that algorithm did the reverse transformation.

andyp1per commented:
Hey @joelucid - I'm beginning to scratch on this to see if I could implement bi-dir dshot for ArduPilot and feel I am lost in a twisty maze of passages all alike -there seem to be a lot of compile time options that have evolved over time. Any chance you could give me a pointer as to where to look? It seems that pwm_output_dshot* are where all the entry points are, but not sure where USE_DSHOT_DMAR, USE_FULL_LL_DRIVER, bitbanging and bitbanding fit in all of these. ArduPilot runs at a lower loop rate to BF and we have lot going on, so I suspect bitbanging is going to be too expensive, but it looks like the DMA config without it is complicated. Any suggestions? We already have DMA-based dshot output.

andyp1per commented:
> The comment you referenced does describe the physical layer in the last step. The last bit sequence is just put to the wire.

@joelucid this still doesn't answer the question - are the bits on the wire encoded as high 1 / low 0 (un-inverted?) or as pulses with long pulse 1, short pulse 0? My reading is that it is the former (and your scope trace makes it look that way also), so a decoder needs to read the line level at (DSHOT_BIT_RATE * 5 / 4) Hz in order to get the 21 bits? Also the being ready after 40us do you mean for instance that for DSHOT600 it is 26us frame + 30us gap + 35us rpm + 40us gap? I'd be really grateful if you could clarify this as I have been pouring over the code and reading blogs and github PRs for hours and still am unclear.

And this applies to BLHeli32 as well going forward? Thanks!

andyp1per commented:
Thanks for all the help guys! <sarc/>

Having had to dig into this a lot in order to implement on F103 I wrote a python based decoder that prints out the bits and the signal for an input. 

To answer my question above, the bits are encoded uninverted, so the first thing the ESC does is pull the line low before sending data. This pull down signals the start of the data. There are no start or stop bits in the classic sense of the word.

The data is sent 30us after the dshot pulse has been received

```
#!/usr/bin/env python

"""
script to encode a telemetry value using dshot and gcr and then display in binary and ascii art
see https://github.com/betaflight/betaflight/pull/8554#issuecomment-512507625
"""

import argparse

parser = argparse.ArgumentParser("bdshot_encoder.py")
parser.add_argument('value', help='Value to encode')

args = parser.parse_args()

def dshot_encode(value):
    packet = (value << 1)

    # compute checksum
    csum = 0
    csum_data = packet
    for i in range(3):
        csum ^= csum_data
        csum_data >>= 4

    csum = ~csum
    csum &= 0xF
    packet = (packet << 4) | csum

    return packet

def rll_encode(value):
    old_bit = 0
    rll_value = 0

    for i in range(1, 21):
        if value & 1:
            new_bit = (1 ^ old_bit)
        else:
            new_bit = old_bit
        value >>= 1
        rll_value |= new_bit << i
        old_bit = new_bit
    return rll_value

def gcr_encode(value):
    expo = 0

    if value:
        while value & 1 == 0:
            value >>= 1
            expo = expo + 1
    
    value = (value & 0x1FF) | (expo << 9)
    value = dshot_encode(value)

    nibble_map = [0x19, 0x1b, 0x12, 0x13, 0x1d, 0x15, 0x16, 0x17, 0x1a, 0x09, 0x0a, 0x0b, 0x1e, 0x0d, 0x0e, 0x0f ]

    new_value = 0
    for i in range(4):
        new_value |= (nibble_map[value & 0xF] << ((3-i) * 5))
        value >> 4

    return rll_encode(new_value)

def print_signal(value):
    old_bit = 0
    print("_ ", sep='', end='')
    for i in range(21):
        bit = (value>>(20-i)) & 1
        if bit != old_bit:
            print(' ', sep='', end='')
        if bit:
            print('_', sep='', end='')
        else:
            print(' ', sep='', end='')
        old_bit = bit
    print(" _", sep='', end='')
    print('')

    print(" |", sep='', end='')
    for i in range(21):
        bit = (value>>(20-i)) & 1
        if bit != old_bit:
            print('|', sep='', end='')
        print(' ', sep='', end='')
        old_bit = bit
    print("| ", sep='', end='')
    print('')

    print(" |", sep='', end='')
    for i in range(21):
        bit = (value>>(20-i)) & 1
        if bit != old_bit:
            print('|', sep='', end='')
        if bit == 0:
            print("_", sep='', end='')
        else:
            print(' ', sep='', end='')
        old_bit = bit
    print("| ", sep='', end='')
    print('')

if args.value.startswith("0b"):
    value = int(args.value[2:], 2)
elif args.value.startswith("0x"):
    value = int(args.value[2:], 16)
else:
    value = int(args.value)

encoded_value = gcr_encode(value)
print("0b{:020b}".format(encoded_value))
print_signal(encoded_value)
```
